fix sort performance issues for dropdowns

MaximilianKrambach opened the pull request:
This is an alternative approach to #4398, doing the sorting at a more deeper level (close to the data store).

Maybe we can build on either of these two PR (both exclude each other) to get a fast sorting of dropdown lists etc.

MaximilianKrambach commented:
This doesn't work, as the background `ViewModelStore` is an object and not an array. The only acceptable part here is the generalization of the sorting, but not the place(s) where the sorting occurs.

(Partial?) sorting should be done in `base-repository:getViewModelListObservable()` if (a) `this.sortFunction` is defined and (b) after a delay of `auditTime( >= 1)`.

MaximilianKrambach commented:
@emanuelschuetze Finn and I now have a (hopefully) working and performant version; 
the category sorting and the  in the 'sort motions in category' is fixed here, too.

tsiegleauq commented on client/src/app/core/repositories/motions/motion-repository.service.ts:
```diff
@@ -801,4 +802,25 @@ export class MotionRepositoryService extends BaseAgendaContentObjectRepository<V
+    /**
+     * Triggers an update for the sort function responsible for the default sorting of data items
+     */
+    public setConfigSortFn(): void {
```
Motion default sorting is missing.
The config has the property:
"Sort motions by" which can either be "Identifier" or "Call list"

Should probably look like in the category repo.

MaximilianKrambach commented:
Everything seems ready now, but I'd rather have another review by @FinnStutzenstein to double check we don't mess up some unsorted/sorted getViewModelList() instances

FinnStutzenstein commented on client/src/app/core/repositories/base-repository.ts:
```diff
@@ -112,7 +147,7 @@ export abstract class BaseRepository<V extends BaseViewModel, M extends BaseMode
 
                     // if an domain object we need was added or changed, update viewModelStore
-                    this.getViewModelList().forEach(ownViewModel => {
+                    Object.values(this.viewModelStore).forEach(ownViewModel => {
```
undo

FinnStutzenstein commented on client/src/app/core/repositories/base-repository.ts:
```diff
@@ -205,10 +261,10 @@ export abstract class BaseRepository<V extends BaseViewModel, M extends BaseMode
+     * update the observable of the list. Also updates the sorting of the view model list.
      */
     protected updateViewModelListObservable(): void {
-        this.viewModelListSubject.next(this.getViewModelList());
```
undo

FinnStutzenstein commented on client/src/app/site/motions/components/motion-detail/motion-detail.component.ts:
```diff
@@ -460,25 +460,22 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit,
-            this.submitterObserver.next(unsortedUsers);
-            this.supporterObserver.next(unsortedUsers);
+        // Initial Filling of the Subjects
+        const initialUsers = this.userRepo.getViewModelList();
```
use sorted list

FinnStutzenstein commented on client/src/app/site/motions/components/motion-detail/motion-detail.component.ts:
```diff
@@ -460,25 +460,22 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit,
+        const initialUsers = this.userRepo.getViewModelList();
+        this.submitterObserver = new BehaviorSubject(initialUsers);
+        this.supporterObserver = new BehaviorSubject(initialUsers);
+        this.categoryObserver = new BehaviorSubject(this.categoryRepo.getViewModelList());
```
use sorted

MaximilianKrambach commented:
Should be in an  okayish state now.

Attention for testing:
- initial sorting in list views
- loading of dropdown-heavy detail views (e.g. a motion detail edit that loads lots of users for submitters/Supporters)
- language change should change language-specific sorting (or at least not crash/stall the sorting)

MaximilianKrambach commented:
(rebase just resolved the new conflicts). I'm more or less waiting for some performance testing 'in the wild' with real user data; else, this is mergable as I see it

emanuelschuetze commented:
Still open:
* category sort
* mediafile sort
* if possible in this pr: parent item field should show agenda item (by using getTitle() for topics)

MaximilianKrambach commented:
> * 
> * if possible in this pr: parent item field should show agenda item (by using getTitle() for topics)

This is not part of this PR, it got broken elsewhere (`search-value-selector` or `getTitle()` in ViewItem or somewhere further down the stack)

MaximilianKrambach commented on client/src/app/site/projector/components/projector-list/projector-list.component.html:
```diff
@@ -19,7 +19,7 @@ <h2 translate>Projectors</h2>
     </span>
     <mat-form-field>
-        <mat-select [disabled]="!!editId" [value]="projectors[0].reference_projector_id" (selectionChange)="onSelectReferenceProjector($event)">
+        <mat-select [disabled]="!!editId" [value]="projectors.length ? projectors[0].reference_projector_id : null" (selectionChange)="onSelectReferenceProjector($event)">
```
@FinnStutzenstein This was one of the two travis testing fails that popped up. Please confirm that m change doesn't destroy anything

FinnStutzenstein replied:
Should be ok.

MaximilianKrambach commented:
I fixed some copy-paste errors (that caused nothing to be sorted at all for some views), and added a mediafile default sorting (by title). Also, initial sorting on a list view is done now, even if the client has never opened the view before

MaximilianKrambach commented:
@emanuelschuetze I fixed the first error, and could not reproduce the second one (might have been a continuation of the first error)

emanuelschuetze commented:
Retested:
1. type error is fixed (confirmed)
2. I get no sortby arrow in sort menu afer opening motion list in a private tab (without indexedDB). Expected: arrow for identifier.

FinnStutzenstein commented:
Rebased on the current master.

tsiegleauq commented:
Rebased again, cleaned up, added better naming unified with current base-repo featuers.
Should work fine now, perhaps one or two setSortFunction calls are missing, but that should be pretty much it.

@emanuelschuetze please test
